### PR TITLE
core, babe, state: refactor tests to use in-memory db

### LINKED ISSUE #639: refactor tests to use in-memory database
we should refactor tests to use an in-memory database when possible to reduce badger DB memory consumption overhead.

### PR DESCRIPTION
**Pull request by noot**
## Changes
- update state, core, babe tests to use in-memory db

## Tests:
```
go test ./... -short
```

### Issues:
closes #639

### CODE REVIEW

File: lib/database/memoryDB.go [hunk 1]
```diff
@@ -61,7 +61,7 @@ func (db *MemDatabase) Get(k []byte) ([]byte, error) {
 	if v, ok := db.db[string(k)]; ok {
 		return v, nil
 	}
-	return nil, errors.New("not found")
+	return nil, errors.New("Key not found")
```
Comment (ryanchristo) [inline lib/database/memoryDB.go]: standard error format starts with lowercase :eyes:

File: lib/babe/median_test.go [hunk 2]
```diff
@@ -207,25 +191,18 @@ func TestEstimateCurrentSlot(t *testing.T) {
...
-	addBlocksToState(t, babesession, 100, dbSrv.Block, uint64(time.Now().Unix())-(babesession.config.SlotDuration/10))
+	addBlocksToState(t, babesession, 100, babesession.blockState, uint64(time.Now().Unix())-(babesession.config.SlotDuration/10))
 
 	res, err := babesession.getCurrentSlot()
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	expected := uint64(101)
+	expected := uint64(100)
```
Comment (ryanchristo) [inline lib/babe/median_test.go]: so the memory database resolved this as well? nice!

File: lib/babe/babe_test.go [hunk 3]
```diff
@@ -73,76 +72,29 @@ func createTestSession(t *testing.T, cfg *SessionConfig) *Session {
...
-		cfg.Runtime = rt
-	}
-
-	var err error
-	if cfg.Keypair == nil {
-		cfg.Keypair, err = sr25519.GenerateKeypair()
+	if cfg.BlockState == nil || cfg.StorageState == nil {
+		dbSrv := state.NewService("")
+		dbSrv.UseMemDB()
+		err = dbSrv.Initialize(genesisHeader, trie.NewEmptyTrie(nil))
```
Comment (ryanchristo) [inline lib/babe/babe_test.go]: this is a really nice improvement :100:

File: dot/state/service.go [hunk 4]
```diff
@@ -112,23 +132,22 @@ func initializeBlockTree(db database.Database, genesisHeader *types.Header) erro
 
 // Start initializes the Storage database and the Block database.
 func (s *Service) Start() error {
-	if s.Storage != nil || s.Block != nil || s.Network != nil {
-		return nil
-	}
+	db := s.db
+	if !s.isMemDB {
```
Comment (ryanchristo) [inline dot/state/service.go]: is this being used for testing purposes only at the moment? should we make note in a comment here?

File: dot/state/service.go [hunk 5]
```diff
@@ -44,25 +45,38 @@ func NewService(path string) *Service {
...
-	if err != nil {
-		return err
-	}
-
-	// initialize database
-	db, err := database.NewBadgerDB(datadir)
-	if err != nil {
-		return err
+	var db database.Database
+	if s.isMemDB {
```
Comment (ryanchristo) [inline dot/state/service.go]: is this being used for testing purposes only at the moment? should we make note in a comment here?

File: lib/database/memoryDB.go (hunk 1 above)
Comment (noot) [reply lib/database/memoryDB.go]: this is done to match the badger DB error that's returned

File: lib/babe/median_test.go (hunk 2 above)
Comment (noot) [reply lib/babe/median_test.go]: yeah, it sped up the gets from the DB

File: dot/state/service.go (hunk 4 above)
Comment (noot) [reply dot/state/service.go]: yes, it's only for testing purposes. I will make a note

File: lib/database/memoryDB.go (hunk 1 above)
Comment (ryanchristo) [reply lib/database/memoryDB.go]: maybe a comment so someone is not tempted to update it in the future

**Pull request merged**